README
NAME
    DBIx::Class::Schema::Versioned::Inline - Defined multiple schema
    versions within resultset classes

VERSION
    Version 0.203

SUMMARY
    Schema versioning for DBIx::Class with version information embedded
    inline in the schema definition.

    See "VERSION NUMBERS" below for important information regarding schema
    version numbering.

SYNOPSIS
     package MyApp::Schema;

     use parent 'DBIx::Class::Schema';

     __PACKAGE__->load_components('Schema::Versioned::Inline');

     our $FIRST_VERSION = '0.001';
     our $VERSION = '0.002';

     __PACKAGE__->load_namespaces;

     ...

     package MyApp::Schema::Result::Bar;

     use base 'DBIx::Class::Core';

     __PACKAGE__->table('bars');

     __PACKAGE__->add_columns(
        "bars_id" => {
            data_type => 'integer', is_auto_increment => 1
        },
        "age" => {
            data_type => "integer", is_nullable => 1
        },
        "height" => {
          data_type => "integer", is_nullable => 1,
          versioned => { since => '0.003' }
        },
        "weight" => {
          data_type => "integer", is_nullable => 1,
          versioned => { until => '0.3' }
        },
     );

     __PACKAGE__->set_primary_key('bars_id');

     __PACKAGE__->has_many(
        'foos', 'TestVersion::Schema::Result::Foo',
        'foos_id', { versioned => { until => '0.003' } },
     );

     __PACKAGE__->resultset_attributes( { versioned => { since => '0.002' } } );

     ...

     package MyApp::Schema::Result::Foo;

     use base 'DBIx::Class::Core';

     __PACKAGE__->table('foos');

     __PACKAGE__->add_columns(
        "foos_id" => {
            data_type => 'integer', is_auto_increment => 1
        },
        "age" => {
            data_type => "integer", is_nullable => 1,
            versioned => { since => '0.002' }
        },
        "height" => {
            data_type => "integer", is_nullable => 1,
            versioned => { until => '0.002' }
        },
        "width" => {
            data_type => "integer", is_nullable => 1,
            versioned => {
                since   => '0.002', renamed_from => 'height',
                changes => {
                    '0.0021' => { is_nullable => 0, default_value => 0 }
                },
            }
        },
        "bars_id" => {
            data_type => 'integer', is_foreign_key => 1, is_nullable => 0,
            versioned => { since => '0.002' }
        },
     );

     __PACKAGE__->set_primary_key('foos_id');

     __PACKAGE__->belongs_to(
        'bar',
        'TestVersion::Schema::Result::Bar',
        'bars_id',
        { versioned => { since => '0.002' } },
     );

     __PACKAGE__->resultset_attributes( { versioned => { until => '0.003' } } );

     ...

     package MyApp::Schema::Upgrade;

     use base 'DBIx::Class::Schema::Versioned::Inline::Upgrade';
     use DBIx::Class::Schema::Versioned::Inline::Upgrade qw/before after/;

     before '0.3.3' => sub {
         my $schema = shift;
         $schema->resultset('Foo')->update({ bar => '' });
     };

     after '0.3.3' => sub {
         my $schema = shift;
         # do something else
     };

DESCRIPTION
    This module extends DBIx::Class::Schema::Versioned using simple 'since'
    and 'until' tokens within result classes to specify the schema version
    at which classes and columns were introduced or removed. Column
    since/until definitions are included as part of 'versioned' info in
    add_column(s).

  since
    When a class is added to a schema at a specific schema version version
    then a 'since' attribute must be added to the class which returns the
    version at which the class was added. For example:

     __PACKAGE__->resultset_attributes({ versioned => { since => '0.002' }});

    It is not necessary to add this to the initial version of a class since
    any class without this atribute is assumed to have existed for ever.

    Using 'since' in a column or relationship definition denotes the version
    at which the column/relation was added. For example:

     __PACKAGE__->add_column(
        "age" => {
            data_type => "integer", is_nullable => 1,
            versioned => { since => '0.002' }
        }
     );

    For changes to column_info such as a change of data_type see "changes".

    Note: if the Result containing the column includes a class-level "since"
    then there is no need to add "since" markers for columns created at the
    same version.

    Relationships are handled in the same way as columns:

     __PACKAGE__->belongs_to(
        'bar',
        'MyApp::Schema::Result::Bar',
        'bars_id',
        { versioned => { since => '0.002' } },
     );

  until
    When used as a class attribute this should be the schema version at
    which the class is to be removed. The underlying database table will be
    removed when the schema is upgraded to this version. Example
    definitions:

     __PACKAGE__->resultset_attributes({ versioned => { until => '0.7' }});

     __PACKAGE__->add_column(
        "age" => {
            data_type => "integer", is_nullable => 1,
            versioned => { until => '0.5' }
        }
     );

    Using 'until' in a column or relationship definition will cause removal
    of the column/relation from the table when the schema is upgraded to
    this version.

  renamed_from
    This is always used alongside 'since' in the renamed class/column and
    there must also be a corresponding 'until' on the old class/column.

    NOTE: when renaming a class the 'renamed_from' value is the table name
    of the old class and NOT the class name.

    For example when renaming a class:

     package MyApp::Schema::Result::Foo;

     __PACKAGE__->table('foos');
     __PACKAGE__->resultset_attributes({ versioned => { until => '0.5 }});

     package MyApp::Schema::Result::Fooey;

     __PACKAGE__->table('fooeys');
     __PACKAGE__->resultset_attributes({
        versioned => { since => '0.5, renamed_from => 'foos' }
     });

    And when renaming a column:

     __PACKAGE__->add_columns(
        "height" => {
            data_type => "integer",
            versioned => { until => '0.002' }
        },
        "width" => {
            data_type => "integer", is_nullable => 0,
            versioned => { since => '0.002', renamed_from => 'height' }
        },
     );

    As can been seen in the example it is possible to modify column
    definitions at the same time as a rename but care should be taken to
    ensure that any data modification (such as ensuring there are no longer
    null values when is_nullable => 0 is introduced) must be handled via
    "Upgrade.pm".

    NOTE: if columns are renamed at the same version that a class/table is
    renamed (for example a renamed PK) then you MUST also add "renamed_from"
    to the column as otherwise data from that column will be lost. In this
    special situation adding "since" to the column is not required.

  changes
    Column definition changes are handled using the "changes" token. A
    hashref is created for each version where the column definition changes
    which details the new column definition in effect from that change
    revision. For example:

     __PACKAGE__->add_columns(
        "item_weight",
        {
            data_type => "integer", is_nullable => 1, default_value => 4,
            versioned => { until => '0.001 },
        },
        "weight",
        {
            data_type => "integer", is_nullable => 1,
            versioned => {
                since        => '0.002',
                renamed_from => 'item_weight',
                changes => {
                    '0.4' => {
                        data_type   => "numeric",
                        size        => [10,2],
                        is_nullable => 1,
                    }
                    '0.401' => {
                        data_type   => "numeric",
                        size        => [10,2],
                        is_nullable => 0,
                        default_value => "0.0",
                    }
                }
            }
        }
     );

    Note: the initial column definition should never be changed since that
    is the definition to be used from when the column is first created until
    the first change is effected.

  Upgrade.pm
    For details on how to apply data modifications that might be required
    during an upgrade see DBIx::Class::Schema::Versioned::Inline::Upgrade.

VERSION NUMBERS
    Under the hood all version numbers are handled using Perl::Version which
    can lead to confusion if you do not understand how Perl versions are
    manipulated. For example:

      $a = Perl::Version->new(0.4)
      $b = Perl::Version->new(0.3)
      $a > $b                           # TRUE

    But things can start to look very odd as soon as we use different
    numbers of decimal places:

      $a = Perl::Version->new(0.12)
      $b = Perl::Version->new(0.30)
      $a > $b                           # TRUE

    And just to add to potential confusion:

      $a = Perl::Version->new(0.12)
      $b = Perl::Version->new("0.30")
      $a > $b                           # FALSE

    The motto of this story is that you must be careful how you manage your
    versions. Please read Perl::Version pod carefully and make sure you
    understand how it operates. To avoid unexpected behaviour it is
    recommended that you always quote the version and if possible use a
    dotted-decimal with at least three components or use simple cardinal
    numbers which can never be confused.

ATTRIBUTES
  schema_versions
    A Set::Equivalence set of PerVersion objects containing all of the
    available schema versions.

    Versions should be added using "add_version".

METHODS
    Many methods are inherited or overloaded from
    DBIx::Class::Schema::Versioned.

  add_version( $version [, $v2, ... ] )
    Adds one or more versions to "schema_versions" set. Arguments can either
    be PerlVersion objects or simple scalars which will be coerced into
    such.

  connection
    Overloaded method. This checks the DBIC schema version against the DB
    version and uses the DB version if it exists or the schema version if
    the database is currently unversioned.

  deploy
    Inherited method. Same as "deploy" in DBIx::Class::Schema but also calls
    "install".

  downgrade
    Call this to attempt to downgrade your database from the version it is
    at to the version this DBIC schema is at. If they are the same it does
    nothing.

  downgrade_single_step
  install
    Inherited method. Call this to initialise a previously unversioned
    database.

  get_db_version
    Override "get_db_version" in DBIx::Class::Schema::Versioned to return
    the version as a PerlVersion object.

  ordered_schema_versions
      $self->ordered_schema_version('desc');

    Optional argument defines the order (ascending or descending). With no
    arg (or an arg we cannot determine direction from) results in ascending.

  schema_first_version
    Returns the current schema class' $FIRST_VERSION in a normalised way.

    If the schema does not define $FIRST_VERSION then all resultsets must
    specify the version at which they were added using "since".

  schema_version
    Override "schema_version" in DBIx::Class::Schema to return the version
    as a PerlVersion object.

  stringified_ordered_schema_versions
    Calls "ordered_schema_versions" with the same args and converts the
    returned list elements to stringified versions.

  upgrade
    Inherited method. Call this to attempt to upgrade your database from the
    version it is at to the version this DBIC schema is at. If they are the
    same it does nothing.

  upgrade_single_step
    Arguments: db_version - the version currently within the db
    Arguments: target_version - the version to upgrade to

    Overloaded method. Call this to attempt to upgrade your database from
    the *db_version* to the *target_version*. If they are the same it does
    nothing.

    All upgrade operations within this step are performed inside a single
    transaction so either all succeed or all fail. If successful the
    dbix_class_schema_versions table is updated with the *target_version*.

    This method may be called repeatedly by the "upgrade" method to upgrade
    through a series of updates.

  versioned_schema
    Arguments: version - the schema version we want to deploy

    Parse schema and remove classes, columns and relationships that are not
    valid for the requested version.

CANDY
    See DBIx::Class::Schema::Versioned::Inline::Candy.

CAVEATS
    Please anticipate API changes in this early state of development.

TODO
    *   Sequence renaming in Pg, MySQL (maybe?). Not required for SQLite.

    *   Index renaming for auto-created indexes for UCs, etc - Pg + others?

    *   Downgrades

    *   Schema validation

AUTHOR
    Peter Mottram (SysPete), "peter@sysnix.com"

CONTRIBUTORS
    Slaven Rezić (eserte) Stefan Hornburg (racke) Peter Rabbitson
    (ribasushi)

BUGS
    This is BETA software so bugs and missing features are expected.

    Please report any bugs or feature requests via the project's GitHub
    issue tracker:

    <https://github.com/Sysnix/dbix-class-schema-versioned-inline/issues>

    I will be notified, and then you'll automatically be notified of
    progress on your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc DBIx::Class::Schema::Versioned::Inline

    You can also look for information at:

    *   GitHub repository

        <https://github.com/Sysnix/dbix-class-schema-versioned-inline>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/DBIx-Class-Schema-Versioned-Inline>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/DBIx-Class-Schema-Versioned-Inline>

    *   Search CPAN

        <http://search.cpan.org/dist/DBIx-Class-Schema-Versioned-Inline/>

ACKNOWLEDGEMENTS
    Thanks to Best Practical Solutions for the Jifty framework and
    Jifty::DBI which inspired this distribution. Many thanks to all of the
    DBIx::Class and SQL::Translator developers for those excellent
    distributions and especially to ribasushi and ilmari for all of their
    help and input.

LICENSE AND COPYRIGHT
    Copyright 2014 Peter Mottram (SysPete).

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for DBIx::Class::Schema::Versioned::Inline

0.203 Fri May 22 17:46:09 CEST 2015

    [TESTS]

    * improve robustness of tmpdir/file usage in SQLite tests so tests pass
      on Windows
    * first connect to schema wrapped in lives_ok in tests

0.201 Wed May 20 12:26:48 CEST 2015

    [TESTS]

    * Improve robustness of MySQL and PostgreSQL smoker tests

0.200 Tue May 19 10:14:04 CEST 2015

    [ENHANCEMENTS]

    * Start using Safe::Isa and Types::PerlVersion

    [BUGS]

    * all version comparisons now use Types::PerlVersion coercions to ensure
      that we handle all version numbers in a uniform fashion.

    [DOCUMENTATION]

    * Add VERSION NUMBERS section to explain the potential dangers with mixed
      version formats.

0.025 Sun Oct 26 17:18:28 CET 2014

    [TESTS]

    * use Class::Load in db engine tests
    * move skip_all into sub BUILD
    * add quote_names => 1 to connect_info for all DBs
    * add diag on failure to unlink sqlite temp file

0.023 Mon Sep  8 17:58:48 CEST 2014

    [TESTS]

    * try to make ::Candy less likely to croak when DBIx::Class::Candy
      is not found
    * reduce test diag verbosity

0.022 Thu Sep  4 14:54:38 CEST 2014

    [ENHANCEMENTS]

    * add $FIRST_VERSION plus schema_first_version method

    [TESTS]

    * hide some useless Test::PostgreSQL warnings
    * add --no-locale to Pg initdb_args to try and prevent failures

0.021 Thu Sep  4 10:34:54 CEST 2014

    * Fix ::Candy tests to prevent smoker failures
    * Fix PostgreSQL tests to prevent failure due to LANG ENV problems

0.020 Wed Sep  3 17:14:11 CEST 2014

    * New SQL::Translator release allows upgrades to work for SQLite and Pg.
      Many thanks to ilmari for assistance in getting patches into SQLT.
    * Start on TODO list in pod.
    * Add Candy component along with tests and required changes.

0.011 Mon Aug  4 14:14:39 CEST 2014

    * ignore versions in Upgrade.pm that do not exist in schema definition
    * better handling of connect_info in upgrade_single_step
    * move sqlt extra handling into upgrade_single_step
    * fix tests for DBIC > 0.08270 (thanks to ribasushi)
    * Pg & SQLite don't like mixed DDL and data changes in since trans
    * fix class has since but missing on renamed PK
    * improved role-based testing

0.010 Wed Jul  9 11:39:27 CEST 2014

    * Upgrade.pm 'after' now implemented with basic tests

0.004 Tue Jul  8 13:16:01 CEST 2014

    * rename_from working for classes (though bugs in SQLT mean that it does
      not work for SQLite and Pg)
    * SKIP failing tests

0.003 Mon Jul  7 19:47:09 CEST 2014

    * Upgrade.pm 'before' now implemented with basic tests

0.002 Thu Jul  3 10:19:43 CEST 2014

    * Add EXLOCK => 0 to File::Temp creation in upgrade-sqlite.t to prevent
      BSD/Mac locking (Slaven Rezić (eserte)) fixes GH#1

0.001 Wed Jul  2 19:08:12 CEST 2014

    * Initial release with missing functionality (Upgrade.pm)



Makefile.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;


WriteMakefile(
    NAME               => 'DBIx::Class::Schema::Versioned::Inline',
    AUTHOR             => q{Peter Mottram (SysPete) <peter@sysnix.com>},
    VERSION_FROM       => 'lib/DBIx/Class/Schema/Versioned/Inline.pm',
    ABSTRACT_FROM      => 'lib/DBIx/Class/Schema/Versioned/Inline.pm',
    LICENSE            => 'perl',
    PL_FILES           => {},
    MIN_PERL_VERSION   => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Class::Load'                 => 0,
        'DBD::SQLite'                 => '1.40',
        'DBIx::Class::Schema::Loader' => '0.07040',
        'File::Spec'                  => '3.47',
        'Test::Deep'                  => 0,
        'Test::Exception'             => 0,
        'Test::Roo'                   => '1.003',
    },
    PREREQ_PM => {
        'DBIx::Class'        => '0.082820',
        'Safe::Isa'          => 0,
        'Set::Equivalence'   => '0.003',
        'Try::Tiny'          => '0.18',
        'Types::PerlVersion' => '0.002',
        'SQL::Translator'    => '0.11021',
    },
    dist => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean      => { FILES => 'DBIx-Class-Schema-Versioned-Inline-*' },
    META_MERGE => {
        resources => {
#<<<
            repository => {
                type => 'git',
                url  => 'git://github.com/Sysnix/dbix-class-schema-versioned-inline.git',
                web  => 'https://github.com/Sysnix/dbix-class-schema-versioned-inline',
            },
            bugtracker => {
                web => 'https://github.com/Sysnix/dbix-class-schema-versioned-inline/issues',
            },
            IRC => 'irc://irc.perl.org/#dbix-class',
#>>>
        },
    },
);


lib/DBIx/Class/Schema/Versioned/Jiftyesque/Result.pm
package DBIx::Class::Schema::Versioned::Jiftyesque::Result;

=head1 NAME

DBIx::Class::Schema::Versioned::Jiftyesque::Result

=cut

use strict;
use warnings;

use base qw(DBIx::Class::Core);

use version 0.77;

use namespace::clean;

sub __UNUSED_register_column {
    my ( $self, $column, $info, @rest ) = @_;

    # this SUCKS and is only good if we are something like:
    #   MyApp::Schema::Result::Foo
    # and also the schema is:
    #   MyApp::Schema
    ( my $schema_class = $self ) =~ s/::Result::.*//;

    # get the schema version cleanly (as per schema_version in DBICS::Versioned)
    my $schema_version;
    {
        no strict 'refs';
        $schema_version = ${"${schema_class}::VERSION"};
    }

    if ( $info->{extra}->{since}
        && version->parse( $info->{extra}->{since} ) >
        version->parse($schema_version) )
    {

        # we don't want this column yet

        $self->remove_column($column);
        return;
    }

    if ( $info->{extra}->{until}
        && version->parse( $info->{extra}->{until} ) <
        version->parse($schema_version) )
    {

        # we don't want this column anymore

        $self->remove_column($column);
        return;
    }

    # register this column
    $self->next::method( $column, $info, @rest );
}

1;


lib/DBIx/Class/Schema/Versioned/Jiftyesque.pm
package DBIx::Class::Schema::Versioned::Jiftyesque;

=head1 NAME

DBIx::Class::Schema::Versioned::Jiftyesque - since/until schema versioning for DBIx::Class in the style of L<Jifty>

=head1 VERSION

Version 0.001

=cut

our $VERSION = '0.001';

=head1 SYNOPSIS

 package MyApp::Schema;

 use base 'DBIx::Class::Schema::Versioned::Jiftyesque';

 our $VERSION = '0.002';

 __PACKAGE__->load_namespaces;

 ...

 package MyApp::Schema::Result::Foo;

 use base 'DBIx::Class::Core';

 __PACKAGE__->table('foos');

 __PACKAGE__->add_columns(
   "foos_id",
   { data_type => 'integer', is_auto_increment => 1 },
   "age",
   { data_type => "integer", is_nullable => 1, extra => { since => '0.002' } },
   "height",
   { data_type => "integer", is_nullable => 1 },
   "bars_id",
   { data_type => 'integer', is_foreign_key => 1, is_nullable => 0, extra => { since => '0.002' } },
 );

 __PACKAGE__->set_primary_key('foos_id');

 __PACKAGE__->has_one(
   'Bar',
   'TestVersion::Schema::Result::Bar',
   'bars_id',
   { extra => { since => '0.002' }},
 );

 __PACKAGE__->resultset_attributes({ extra => { until => '0.002' }});

 ...

 package MyApp::Schema::Result::Bar;

 use base 'DBIx::Class::Core';

 __PACKAGE__->table('bars');

 __PACKAGE__->add_columns(
   "bars_id",
   { data_type => 'integer', is_auto_increment => 1, },
   "age",
   { data_type => "integer", is_nullable => 1 },
   "height",
   { data_type => "integer", is_nullable => 1, extra => { until => '0.003' } },
   "weight",
   { data_type => "integer", is_nullable => 1, extra => { until => '0.3' } },
 );

 __PACKAGE__->set_primary_key('bars_id');

 __PACKAGE__->belongs_to(
   'Foo',
   'TestVersion::Schema::Result::Foo',
   'bars_id',
   { extra => { until => '0.002' }},
 );

 __PACKAGE__->resultset_attributes({ extra => { since => '0.002' }});

=head1 DESCRIPTION

This module extends L<DBIx::Class::Schema::Versioned> using simple 'since' and 'until' markers within result classes to specify the schema version at which classes and columns were introduced or removed. Column since/until definitions are included as part of 'extra' info in add_column(s).

=head2 since

When a class is added to a schema at a specific schema version version then a 'since' attribute must be added to the class which returns the version at which the class was added. For example:

 __PACKAGE__->resultset_attributes({ extra => { since => '0.002' }});

It is not necessary to add this to the initial version of a class since any class without this atribute is assumed to have existed for ever.

Using 'since' in a column or relationship definition denotes the version at which the column/relation was added.

=head2 until

When used as a class attribute this should be the final schema version at which the class is to be used. The underlying database table will be removed when the schema is upgraded to a higher version. Example definition:

 __PACKAGE__->resultset_attributes({ extra => { since => '0.3' }});

Using 'until' in a column or relationship definition will cause removal of the column/relation from the table when the schema is upgraded past this version.

=head2 renamed_from

For renaming a class:

 package MyApp::Schema::Result::Foo;

 __PACKAGE__->table('foos');
 __PACKAGE__->resultset_attributes({ extra => { until => '0.4 }});

 package MyApp::Schema::Result::Fooey;

 __PACKAGE__->table('fooeys');
 __PACKAGE__->resultset_attributes({
     extra => { since => '0.5, renamed_from => 'Foo' }
 });

Or for renaming a column:

 __PACKAGE__->add_columns(
     "height",
     { data_type => "integer", extra => { until => '0.001' } },
     "width",
     { data_type => "integer", extra => {
         since => '0.002', renamed_from => 'height' }
     },
 )

=cut

use warnings;
use strict;

use base 'DBIx::Class::Schema::Versioned';

use Carp;
use Data::Dumper::Concise;
use SQL::Translator;
use SQL::Translator::Diff;
use version 0.77;

our @schema_versions;

=head1 METHODS

Many methods are inherited or overloaded from L<DBIx::Class::Schema::Versioned>.

=head2 connection

Overloaded method. This checks the DBIC schema version against the DB version and ???? TODO ???? (undecided so far) if they are not the same or if the DB is unversioned.

=cut

sub connection {
    my $self = shift;
    $self->next::method(@_);

    my $conn_info = $self->storage->connect_info;
    $self->{vschema} = DBIx::Class::Version->connect(@$conn_info);
    my $conn_attrs = $self->{vschema}->storage->_dbic_connect_attributes || {};

    my $vtable = $self->{vschema}->resultset('Table');

    my $version = $conn_attrs->{_version} || $self->get_db_version();

    unless ($version) {

        # TODO: checks for unversioned
        # - do we throw exception?
        # - do we install automatically?
        # - can we be passed some method or connect arg?
        # for now just set $pversion to schema_version
        $version = $self->schema_version;
    }

    $self->versioned_schema( $version, $conn_attrs->{_type} );

    return $self;
}

=head2 deploy

Inherited method. Same as L<DBIx::Class::Schema/deploy> but also calls C<install>.

=head2 install

Inherited method. Call this to initialise a previously unversioned database.

=head2 ordered_schema_versions

Overloaded method. Returns an ordered list of schema versions. This is then used to produce a set of steps to upgrade through to achieve the required schema version.

=cut

sub ordered_schema_versions {
    my $self = shift;

    # add schema and database versions to list
    push @schema_versions, $self->get_db_version, $self->schema_version;

    # add Upgrade versions
    my $upgradeclass = ref($self) . "::Upgrade";
    eval {
        eval "require $upgradeclass" or return;
        push @schema_versions, $upgradeclass->versions;
    };

    return sort { version->parse($a) <=> version->parse($b) } do {
        my %seen;
        grep { !$seen{$_}++ } @schema_versions;
    };
}

=head2 upgrade

Inherited method. Call this to attempt to upgrade your database from the version it is at to the version this DBIC schema is at. If they are the same it does nothing.

=head2 upgrade_single_step

=over 4
 
=item Arguments: db_version - the version currently within the db
 
=item Arguments: target_version - the version to upgrade to

=back

Overloaded method. Call this to attempt to upgrade your database from the I<db_version> to the I<target_version>. If they are the same it does nothing.

All upgrade operations within this step are performed inside a single transaction so either all succeed or all fail. If successful the dbix_class_schema_versions table is updated with the I<target_version>.

This method may be called repeatedly by the L</upgrade> method to upgrade through a series of updates.

=cut

sub upgrade_single_step {
    my ( $self, $db_version, $target_version ) = @_;

    # db and schema at same version. do nothing
    if ( $db_version eq $target_version ) {
        carp 'Upgrade not necessary';
        return;
    }
    carp "attempting upgrade from $db_version to $target_version";

    my $sqlt_type = $self->storage->sqlt_type;

    # add Upgrade versions
    my $upgradeclass = ref($self) . "::Upgrade";
    eval {
        eval "require $upgradeclass" or return;
        my @sql = $upgradeclass->upgrade_to($target_version);
    warn Dumper(@sql);
    };

    # translate current schema

    my $curr_tr = SQL::Translator->new(
        no_comments => 1,
        parser   => 'SQL::Translator::Parser::DBIx::Class',
        #parser   => 'SQL::Translator::Parser::DBIx::Class::Jiftyesque',
        parser_args => {
            dbic_schema => $self,
        },
        producer => $sqlt_type,
        show_warnings => 1,
    ) or $self->throw_exception(SQL::Translator->error);
    $curr_tr->translate;

    # translate target schema

    # our target future-versioned connect causes warning messages we don't want
    my $old_DBIC_NO_VERSION_CHECK = $ENV{DBIC_NO_VERSION_CHECK} || 0;
    $ENV{DBIC_NO_VERSION_CHECK} = 1;

    # we'll reuse connect_info from existing schema for target ver connect
    my $connect_info = $self->storage->connect_info;

    # padd out user/pass if they don't exist
    while ( scalar @$connect_info < 3 ) {
        push @$connect_info, undef;
    }

    # drop anything extra
    while ( scalar @$connect_info > 3 ) {
        pop @$connect_info;
    }

    # add next version
    push @$connect_info, { _version => $target_version };

    my $target_schema = ref($self)->connect( @$connect_info );

    # turn noises back to normal level
    $ENV{DBIC_NO_VERSION_CHECK} = $old_DBIC_NO_VERSION_CHECK;

    my $target_tr = SQL::Translator->new(
        no_comments => 1,
        #parser   => 'SQL::Translator::Parser::DBIx::Class::Jiftyesque',
        parser   => 'SQL::Translator::Parser::DBIx::Class',
        parser_args => {
            dbic_schema => $target_schema,
        },
        producer => $sqlt_type,
        show_warnings => 1,
    ) or $self->throw_exception(SQL::Translator->error);
    $target_tr->translate;

    my @diff = SQL::Translator::Diff->new({
        output_db     => $sqlt_type,
        source_schema => $curr_tr->schema,
        target_schema => $target_tr->schema,
        ignore_index_names => 1,
        ignore_constraint_names => 1,
    })->compute_differences->produce_diff_sql;


    foreach my $line (@diff) {
        $self->storage->dbh_do(
            sub {
                my ( $storage, $dbh ) = @_;
                $dbh->do($line);
            }
        );
    }

    # set row in dbix_class_schema_versions table
    $self->_set_db_version({version => $target_version});
}

=head2 versioned_schema

=over 4

=item Arguments: version - the schema version we want to deploy

=back

Parse schema and remove classes, columns and relationships that are not valid for the requested version.

=cut

sub versioned_schema {
    my ( $self, $_version ) = @_;

    my $pversion = version->parse($_version);

    foreach my $source_name ( $self->sources ) {

        my $source = $self->source($source_name);

        # check columns before deciding on class-level since/until to make sure
        # we don't miss any versions

        foreach my $column ( $source->columns ) {

            my $extra = $source->column_info($column)->{extra};

            my $since = $extra->{since};
            my $until = $extra->{until};

            my $name = "$source_name column $column";
            my $sub  = sub {
                my $source = shift;
                $source->remove_column($column);
            };
            $self->_since_until( $pversion, $extra->{since},
                $extra->{until}, $name, $sub, $source );
        }

        # now check relations

        foreach my $relation_name ( $source->relationships ) {

            my $attrs = $source->relationship_info($relation_name)->{attrs};

            next unless defined $attrs;

            my $extra = $attrs->{extra};

            next unless defined $extra;

            my $since = $extra->{since};
            my $until = $extra->{until};

            my $name = "$source_name relationship $relation_name";
            my $sub  = sub {
                my $source = shift;
                my %rels   = %{ $source->_relationships };
                delete $rels{$relation_name};
                $source->_relationships( \%rels );
            };
            $self->_since_until( $pversion, $extra->{since},
                $extra->{until}, $name, $sub, $source );
        }

        # now check class-level since/until

        my ( $since, $until );

        my $extra = $source->resultset_attributes->{extra};

        if ( defined $extra ) {
            $since = $extra->{since} if defined $extra->{since};
            $until = $extra->{until} if defined $extra->{until};
        }

        my $name = $source_name;
        my $sub  = sub {
            my $class = shift;
            $class->unregister_source($source_name);
        };
        $self->_since_until( $pversion, $extra->{since}, $extra->{until},
            $name, $sub, $self );
    }
}

sub _since_until {
    my ( $self, $pversion, $since, $until, $name, $sub, $thing ) = @_;

    push (@schema_versions, $since) if $since;
    push (@schema_versions, $until) if $until;

    if (   $since
        && $until
        && ( version->parse($since) > version->parse($until) ) )
    {
        $self->throw_exception("$name has since greater than until");
    }

    # until is absolute so parse before since
    if ( $until && $pversion > version->parse($until) )
    {
        $sub->($thing);
    }
    if ( $since && $pversion < version->parse($since) )
    {
        $sub->($thing);
    }
}

=head1 CAVEATS

Please anticipate API changes in this early state of development.

=head1 AUTHOR

Peter Mottram (SysPete), "peter@sysnix.com"

=head1 BUGS

LOTS of bugs and missing features right now.

NOTE: upgrades are NOT yet implemented.

Please report any bugs or feature requests via the project's GitHub issue tracker:

L<https://github.com/Sysnix/dbix-class-schema-versioned-jiftyesque/issues>

I will be notified, and then you'll automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc DBIx::Class::Schema::Versioned::Jiftyesque

You can also look for information at:

=over 4

=item * GitHub repository

L<https://github.com/Sysnix/dbix-class-schema-versioned-jiftyesque>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/DBIx-Class-Schema-Versioned-Jiftyesque>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/DBIx-Class-Schema-Versioned-Jiftyesque>

=item * Search CPAN

L<http://search.cpan.org/dist/DBIx-Class-Schema-Versioned-Jiftyesque/>

=back

=head1 ACKNOWLEDGEMENTS

Thanks to Best Practical Solutions for the L<Jifty> framework and L<Jifty::DBI> which inspired this distribution. Thanks also to Matt S. Trout and all of the L<DBIx::Class> developers for an excellent distribution.

=head1 LICENSE AND COPYRIGHT

Copyright 2014 Peter Mottram (SysPete).

This program is free software; you can redistribute it and/or modify it under the terms of either: the GNU General Public License as published by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;
